project/canvas.py

```diff
@@ -0,0 +1,650 @@
+
+            if width > 600:
+                vbar = tk.AsyncScrollbar(self.frame, orient=tk.VERTICAL)
+                vbar.pack(side=tk.RIGHT, fill=tk.Y)
+                vbar.config(command=self.xview)
+                self.config(xscrollcommand=vbar.set)  # intentional switch
+            self.pack(side=tk.LEFT, expand=True, fill=tk.BOTH)
+
+        else:
+            super().__init__(master, height=height, width=width, bg="white")
+            self.pack(side=tk.LEFT)
+
+        self.width, self.height = width, height
+
+        self.pil_image = Image.new("RGB", (width, height), (255, 255, 255))
+        self.pil_draw = ImageDraw.Draw(self.pil_image)
+
+        self.undo_list = []
+        self.redo_list = []
+
+        self._master = master
+
+    async def add_pixel(self, x: int, y: int, colour: Colour):
+        """
+        Adds pixel to both the displayed canvas and the backend PIL image
+
+        Parameters
+        ----------
+        x: int
+            The x coordinate of the pixel
+        y: int
+            The y coordinate of the pixel
+        colour: Colour
+            The fill colour of the pixel
+        """
+
+        await self.create_line(x, y, x, y, fill=colour.as_hex)
+        self.pil_draw.point([(x - 1, y - 1)], fill=colour.as_rgb)
+
+    async def undo(self):
+        """Undoes the most previous action by taking the most recent value from the undo_list"""
+        if self.undo_list:
+            x, y, old_colour = self.undo_list.pop()
+            new_colour = self.pil_image.getpixel((x, y))
+            self.redo_list.append((x, y, Colour.from_rgb(new_colour)))
+            await self.add_pixel(x, y, old_colour)
+
+    async def redo(self):
+        """Redoes the most previous action by taking the most recent value from the redo_list"""
+        if self.redo_list:
+            x, y, new_colour = self.redo_list.pop()
+            old_colour = self.pil_image.getpixel((x, y))
+            self.undo_list.append((x, y, Colour.from_rgb(old_colour)))
+            await self.add_pixel(x, y, new_colour)
+
+    async def save(self, file: typing.Union[str, pathlib.Path]):
+        """Shortcut to save the PIL file"""
+        # Get the file extension
+        ext = str(file).split(".")[-1]
+        buffer = BytesIO()
+
+        # Save into buffer
+        self.pil_image.save(buffer, format=ext)
+
+        # Find out how many bytes the file holds, and reset file pointer
+        max_bytes = buffer.tell()
+        buffer.seek(0)
+
+        async with aiofiles.open(file, "wb") as fp:
+            # For every byte
+            for i in range(max_bytes):
+                current_byte = buffer.read(1)
+                # Open a window
+                root = tk.AsyncToplevel(self._master)
+                # ... that the user cannot kill
+                root.protocol('WM_DELETE_WINDOW', lambda *i: None)
+
+                async def cb(i=i):  # No late binding
```

Comment:
what in the world is this `i=i` supposed to do? o.O

Reply:
The i=i is used to prevent late binding as the variable `i` would change before the callback function is called and hence, one triggered, will use the updated version of the variable. This is equivalent to passing in the argument early via the `functools.partial` method

Reply:
but.. that variable isn't even used in this function.

Reply:
The use of `i=i` here is actually a remnant of the original save dialogue behaviour. During my testing, I had every dialogue open at once, for the user to press in order by themselves, but I felt that to be too harsh and too easy to misclick, which would not fit the brief as well, in my eyes. When I added the `wait_window` I never removed the late binding hack.